Add guidance for sortable table

sarawilcox opened the pull request:
## Description

First stab at guidance for the sortable table. For the sortable table working group. 

Feel free to comment or add to this. 

### Related issue

<!--- Optional: if there is an open GitHub or JIRA issue, please link to the issue here -->

## Checklist

<!-- Ensure each of the points below have been considered and completed where applicable -->

- [ ] Tested against the [NHS.UK frontend testing policy](https://github.com/nhsuk/nhsuk-frontend/blob/main/docs/contributing/testing.md) (Resolution, Browser & Accessibility)
- [ ] Code follows the [NHS.UK frontend coding standards](https://github.com/nhsuk/nhsuk-frontend/blob/main/docs/contributing/coding-standards.md)
- [ ] Version number has been increased in `package.json` (using [SEMVER](https://semver.org/))
- [ ] CHANGELOG entry
- [ ] [Whats new page](https://service-manual.nhs.uk/whats-new) entry

sarawilcox committed:
Add initial draft of guidance for sortable table

sarawilcox committed:
Add initial draft of guidance for sortable table

sarawilcox committed:
Tweak content

sarawilcox committed:
Tweak content

cjforms commented:
First draft so some quick comments:

1. You've got a stray 'should' in this: "<p>[What types of data should not be sorted?]</p>" And to be honest, I'm not entirely clear what the guidance says in this case.
2. Another question: why is is #basic-tables plural when the other types of table are singular in their tags?
3. And another question: what do you suggest people do with data for the general public instead of using a table?
4. I don't really understand "Do not let users sort a table when the data must be viewed as is". What sort of data must be viewed 'as it is'? Is there any example?

anandamaryon1 committed:
add mention of defualt sorting order

anandamaryon1 committed:
add mention of defualt sorting order

anandamaryon1 committed:
Add example placeholder and add suggestion for what not to sort

anandamaryon1 committed:
Add example placeholder and add suggestion for what not to sort

anandamaryon1 committed:
suggestion for when not to use sortable table

anandamaryon1 committed:
suggestion for when not to use sortable table

anandamaryon1 committed:
tweak session storage guidance

anandamaryon1 committed:
tweak session storage guidance

anandamaryon1 committed:
tweak server-side sorting guidance

anandamaryon1 committed:
tweak server-side sorting guidance

sarawilcox committed:
Iterate content

sarawilcox committed:
Iterate content

anandamaryon1 committed:
add example

anandamaryon1 committed:
install preview branch and add example

sarawilcox committed:
Iterate sortable tables content

sarawilcox committed:
Iterate sortable tables content

sarawilcox committed:
Update content re when to use a table

sarawilcox committed:
Update content re when to use a table

anandamaryon1 committed:
update frontend preview

colinrotherham commented:
I've switched to a new preview branch:

```console
npm install --save github:nhsuk/nhsuk-frontend#50ec9a5c8
```

All working again

colinrotherham committed:
Temporarily use NHS.UK frontend preview

sarawilcox committed:
Iterate guidance on sortable tables

sarawilcox committed:
Iterate guidance on sortable tables

sarawilcox commented:
Hi @frankieroberto and @edwardhorsford , would you have a mo to look at the updated guidance please? You can see it here: https://nhsuk-service-manual-pr-2382.herokuapp.com/design-system/components/table#sortable-table.

Is what I've said about data attributes correct? Note: I haven't worked out how to get the code about data-sort-values to display in the browser, for example, where it says 30 November 2026. 

Frankie, you said in Slack that the default is ascending but the example seems to me to be descending. Is that right? 

Are we going to allow people to change the default? 

Is what I've written about prefixes "a" and "the" OK? I think people may ask how they do that? Is there anything we can say?

frankieroberto commented:
> Frankie, you said in Slack that the default is ascending but the example seems to me to be descending. Is that right?
> 
> Are we going to allow people to change the default?

It’s defaulting to ascending for me (when you click a new column heading)?  I can’t remember why we picked that as the default though! (I guess you’re more often interested in seeing the biggest numbers?) It’s not yet configurable, but I’ll work on that.

frankieroberto committed:
Merge branch 'main' into sortable-tables

frankieroberto committed:
Fix merge

frankieroberto committed:
Downgrade slug

frankieroberto committed:
Force slug to 9.1.0

Temporary fix

frankieroberto committed:
see if this fixes dependency issue

frankieroberto committed:
Add an example

frankieroberto commented:
@sarawilcox @edwardhorsford I’ve had another go at updating the guidance for changing the custom sort order. It now includes another example showing 3 columns with different types of `data-sort-value` values. I think this is probably needed, as otherwise it’s not super clear how to add these to the HTML or Nunjucks?

Feel free to edit - I might have made it too long?!

sarawilcox commented:
I like that, @frankieroberto . I think it's much clearer. Thanks for doing it. :blue_heart:

The only thing I'm not clear about (it may be that I'm misunderstanding something) is the bit about default order. In our UR we saw other people get confused about this too so it would be good to make it as clear as poss.

When you come to the page, the childhood vaccination table looks like this, which to me suggests that the default is descending - the MMR column numbers go from large to small and the header shows a down arrow. I assumed that the default example (usually the first on the page) would use default order.

<img width="871" height="589" alt="Screenshot 2026-04-08 at 10 38 22" src="https://github.com/user-attachments/assets/400dbf2b-4575-4762-b6e4-6a99dbeb6401" />

I'm also now wondering why the nations are in the order they're in? Would they be better either as normally quoted: English, Scotland, Wales, Northern Ireland or in alphabetical order? 

With the **Organisations using this service** table, the first column (the nations) shows an up arrow, which says to me ascending, with alphabetical order A to Z. Have I got that right?

<img width="455" height="413" alt="Screenshot 2026-04-08 at 10 34 44" src="https://github.com/user-attachments/assets/5f2dde5b-5fe5-46b0-8eef-5327b67839db" />

sarawilcox commented:
I think this is almost ready to go, but we're waiting to see if the pilot testing brings up any other issues.

frankieroberto committed:
Add details of using data-initial-sort

frankieroberto committed:
Add details of using data-initial-sort

frankieroberto commented:
@sarawilcox ah sorry, there are 2 different meanings of 'initial sort' being described here:

1) the sort order for the table when the page is first loaded.
2) what happens when you click the header for a currently-unsorted column.

For that second scenario, the default is 'ascending', so it'll go from unsorted to ascending, and then if you click it again it'll keep toggling between the 2.

However on @edwardhorsford's suggestion I’ve added an option to change this default on a per-column basis by setting `data-initial-sort="descending"`. This might make sense, for example, if you think that most users will be more interested in seeing the biggest numbers first for that column. They can still switch to smallest-first by clicking again, so all it’s doing is saving a click.


On the first question - what the default sort order should be when the page first loads, that’s up to the user and there’s no default set (they can choose it by setting `aria-sort="ascending"` or `aria-sort="descending"` on one of the columns. It probably depends on the context?


Tricky thing to explain all this though! I’ve done a quick update to try and explain the new `data-initial-sort` option (which isn't working in the preview yet) - does it make sense at all?

frankieroberto commented:
> I'm also now wondering why the nations are in the order they're in? Would they be better either as normally quoted: English, Scotland, Wales, Northern Ireland or in alphabetical order?

At the moment in the example the initial sort order is descending by MMR vaccination rates, so they’re in that order. I’d made the first column un-sortable to try and include an example of where not all columns are sortable - but maybe you might well want to sort by region name, and so we should find another example? 🤔 

> With the Organisations using this service table, the first column (the nations) shows an up arrow, which says to me ascending, with alphabetical order A to Z. Have I got that right?

Yep, they’re in ascending order by name. "The Christie" comes first as it's sorting on "Christie" with the "the" dropped (as an example of that feature). Does that help, or is it too confusing?

edwardhorsford commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,11 +71,96 @@
+    <li>text in alphabetical order A to Z</li>
+  </ul>
+  <p>Descending order (down arrow) reverses this.</p>
+  <p>The default is ascending first, but you can change the default.</p>
```
```suggestion
  <p>The default is ascending first, but you can change the default per column.</p>
```

edwardhorsford commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,11 +71,96 @@
+  <p>Do not use a responsive table if your content is easier to read and understand on a small screen in a non-responsive layout. Use a <a href="#basic-tables">basic table</a> instead.</p>
+
+  <h2 id="sortable-table">Sortable table</h2>
+  <p>A sortable table lets users sort columns in ascending or descending order.</p>
```
I wonder if it's worth mentioning this is a progressive enhancement / requires js. / teams still need to provide the initial data with a suitable sort.

edwardhorsford commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,11 +71,96 @@
+  <p>A sortable table lets users sort columns in ascending or descending order.</p>
+  <p>Ascending order (up arrow) arranges:</p>
+  <ul>
+    <li>numerical data from the lowest to the highest value</li>
+    <li>text in alphabetical order A to Z</li>
```
@frankieroberto how does it handle numbers as strings?

frankieroberto replied:
@edwardhorsford If the string contains only a number (ie javascript can interpret it using `Number()` - which means `"10.2"` and `"-5"` both work) then it's handled as a number.

If there are any other characters present, then it's treated as text - eg `"£100"` or `"10ml"` or `"USD 10"`. For those scenarios, you should provide the number itself in the `data-sort-value` attribute.

Hopefully the guidance and the example makes this clear, but if not we should make it clearer!

colinrotherham replied:
Just to clarify this now we're using [`Intl.Collator` sorting](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Collator)

You can now:

1. Set `format: "numeric"` and let the browser sort units (`£`, `ml`, `%`) automatically
3. Set `sortValue` for different units only, e.g. `sortValue: "£0.50"` for `text: "50p"`
3. Set `sortValue` on every cell in the sortable column

It's quite hard to explain so 3) is probably easier

edwardhorsford commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,11 +71,96 @@
+  }) }}
+
+  <h3 id="when-to-use-a-sortable-table">When to use a sortable table</h3>
+  <p>Use this table when users need to find and prioritise data, for example in staff systems with large data sets.</p>
```
This is only client side, right? so likely without pagination, and likely not large data sets? Not sure 'large data set' needs to be mentioned.

frankieroberto replied:
The component does actually support server-side sorting now as an alternative, although only via full page reloads using link tags rather than ajax.

You use the same syntax, but add an `href`:

```
head: {
 items: [
   {
     text: "Name",
     href: "?sort=name",
     attributes: {
       "aria-sort": "ascending"
     } 
   },
   {
     text: "Date",
     href: "?sort=date",
     attributes: {
       "aria-sort": "none"
     } 
   }
 ]
}
```

However the guidance doesn’t currently mention this.

The client-side sorting is still a progressive enhancement of a non-sortable table. The server-side sorting is for where you have pagination.

edwardhorsford commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,11 +71,96 @@
+    <li>how much data to display at once</li>
+    <li>how it should be ordered by default</li>
+    <li>whether to place unsortable data such as empty or null values at the start or end of the table</li>
+  </ul>
```
This feels a touch detailed for me. I could imagine teams should consider if it's suitable to sort, but I doubt a team should would go in to this much detail for each table of data.

edwardhorsford commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,11 +71,96 @@
+    <li>whether to place unsortable data such as empty or null values at the start or end of the table</li>
+  </ul>
+
+  <p>Do not allow sorting on columns that cannot be meaningfully ordered, for example free text.</p>
```
Unsure that free text is implicitly unsortable - you can still sort alphabetically and that may be correct.

frankieroberto replied:
"Free text" is perhaps jargon too?

Maybe a better examples would be "a random id number" (like NHS number?) or "a comment from a user" or similar?

edwardhorsford replied:
Sorting by NHS number also seems like something a team may well want to do.

frankieroberto replied:
@edwardhorsford oh really? I guess it might be useful in contexts where you're comparing 2 lists...

Hard to think of a good example of where it's unlikely it’d make sense to sort it!

edwardhorsford commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,11 +71,96 @@
+
+  <p>Do not sort data alphabetically if it has an established logical sequence, for example months of the year, January to December.</p>
+
+  <p>You may want to remove common prefixes like "a" and "the", treating them as "stop words", so that "The Whittington Hospital" comes under W, not under T.</p>
```
I wonder if this should mention that you can do this without affecting what's visually shown to the user. I know this is somewhat covered in the next sentence, but I'm not sure that it's obvious that's the case - they seem like two distinct things as written.

Perhaps this sentence could run on _after_ the one about sort order.

edwardhorsford commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,11 +71,96 @@
+    <li>prefixes: The Whittington Hospital would be <code>Whittington Hospital</code></li>
+  </ul>
+
+  <p>By default when you click on the column heading for an unsorted column, it will sort that column in ascending order. You can change this by adding <code>data-initial-sort="descending"</code> to the column heading if descending is the more useful initial sort.</p>
```
```suggestion
  <p>By default when you click on the column heading for an unsorted column, it will sort that column in ascending order. You can change this by adding <code>data-initial-sort="descending"</code> to the column heading if descending is the more useful initial sort direction.</p>
```

edwardhorsford commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,11 +71,96 @@
+  <p>A page refresh or navigation must return the page to its default sort order, unless you add functionality to allow users to save their sorting order explicitly.</p>
+
+  <h4 id="server-side-sorting">Server-side sorting</h4>
+  <p>This component provides client-side sorting, which only sorts the data currently visible in the table on the page.</p>
```
Should this describe what client side means or that it relies on JavaScript?

edwardhorsford commented:
Something I'm not sure is covered is how null or empty values get sorted by default, or indeed if we have any opinions on how they should be handled. It's briefly mentioned in UR - but how will they be treated currently?

Eg should they go to the end? should you try to sort them based on something else?

On a previous project where I designed sortable tables, we had config to decide whether unsortables (null values) went to start or end, or were always at the end, regardless of the sort direction. 

We also supported a fallback sort column - so in the case of a tie (two values the same) or empty cell, we'd fall back to a different column and sort based on that. This meant that we could have empty values all sorted to the end, but then arranged alphabetically.

With the `data-sort-direction` you can _somewhat do this, though it might be cumbersome. I don't think it would be possible to set that empty cells always sort to the end and then alphabetically.

frankieroberto committed:
Make wording clearer

Co-authored-by: Ed Horsford <mail@edwardhorsford.com>

frankieroberto committed:
Make wording clearer

Co-authored-by: Ed Horsford <mail@edwardhorsford.com>

frankieroberto committed:
Content tweak

Co-authored-by: Ed Horsford <mail@edwardhorsford.com>

frankieroberto committed:
Content tweak

Co-authored-by: Ed Horsford <mail@edwardhorsford.com>

frankieroberto commented:
@edwardhorsford on how empty values should be sorted, currently the javascript sorts blanks first if ascending and numerical, and last if ascending and alphabetic.

This is based on this JavaScript:

```js
return typeof valueA === 'number' && typeof valueB === 'number'
        ? valueA - valueB
        : valueA.toString().localeCompare(valueB.toString())
```        

We’ve not given any thought yet to whether this is right or not, or whether it should be configurable.

The current table guidance also tells people not to leave empty cells at all:

<img width="796" height="325" alt="Screenshot 2026-04-13 at 16 03 11" src="https://github.com/user-attachments/assets/e3867ee8-7f1f-4ac2-9ce7-ad6f9aa9ada5" />

If you use "No data" or "Not applicable" then you’d have to use it alongside a `data-sort-value` attribute, and you could choose to either use `"0"`, or `"-1"` if you wanted it to appear first, or `"999999999999"` or something if you wanted it to appear last (although that's a bit of a hack!).

edwardhorsford commented:
>If you use "No data" or "Not applicable" then you’d have to use it alongside a data-sort-value attribute, and you could choose to either use "0", or "-1" if you wanted it to appear first, or "999999999999" or something if you wanted it to appear last (although that's a bit of a hack!).

I think each of those options only works for _one_ of the sort directions? So `9999999` might work to sort to end - but as soon as you reverse the sort direction they'd be at the start. Probably not a thing for now, but I don't think the current approach would allow someone to decide unsortables were *always* last.

frankieroberto commented:
> I think each of those options only works for _one_ of the sort directions? So `9999999` might work to sort to end - but as soon as you reverse the sort direction they'd be at the start. Probably not a thing for now, but I don't think the current approach would allow someone to decide unsortables were _always_ last.

Ah good point. Yeah that’s currently impossible. If this comes up as a need we could figure out a way to support it in future. Maybe you'd set `data-sort-always-last="true"` or something, instead of a `data-sort-value`? 🤷

sarawilcox committed:
Tweak guidance

sarawilcox committed:
Make content tweaks re sort order

sarawilcox committed:
Iterate guidance

anandamaryon1 committed:
Update examples, inc to use row headers

sarawilcox committed:
Iterate content

anandamaryon1 committed:
add data-initial-sort to sortable table vaccine example

sarawilcox committed:
Tweak wording re sorting across pages

frankieroberto committed:
Add guidance and examples on server-side version

frankieroberto committed:
Update to use new param names

Changed over in nhsuk-frontend

frankieroberto committed:
Switch to using new Nunjucks params

frankieroberto committed:
Restructure guidance using h5s

anandamaryon1 committed:
fix example markup

colinrotherham committed:
Use double quotes

colinrotherham committed:
Workaround for server-side table sorting needing `aria-sort` attributes

anandamaryon1 committed:
add sort to example's nation column, add no data example and guidance

anandamaryon1 committed:
add non-sort ODS column to example

anandamaryon1 committed:
add guidance about default sort order

sarawilcox committed:
iterate guidance for sortable table

anandamaryon1 committed:
add subheadings and reorder sorting content

sarawilcox committed:
iterate content

frankieroberto committed:
Add an example

anandamaryon1 committed:
add guidance and example for scrolling table

anandamaryon1 committed:
add hr's between home page content blocks

anandamaryon1 committed:
tweak scrolling table guidance

anandamaryon1 committed:
tweak responsive table wording

colinrotherham committed:
Add JavaScript toggles to sortable table examples

sarawilcox committed:
Do what's new and changelog

sarawilcox committed:
change What's coming soon to Coming soon

anandamaryon1 committed:
Tweak wording for backlog CTA

anandamaryon1 committed:
switch coming soon bullets to summary list

sarawilcox committed:
tweak wording on coming soon and backlog page

anandamaryon1 committed:
Add guidance and example for compact tables

anandamaryon1 committed:
move compact, explain combining columns

anandamaryon1 committed:
add jump links

sarawilcox committed:
tweak sortable table content

colinrotherham committed:
Formatting

colinrotherham committed:
Enable compact table

colinrotherham committed:
Describe Nunjucks `compact: true` option alongside modifier class

frankieroberto committed:
Remove redundant macro attributes

These aren't needed for server-side table sorting.

sarawilcox committed:
Add sentence on tables page re server-side example

frankieroberto commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,11 +83,110 @@
+
+  <p>If your table is paginated, do not use the JavaScript-based sorting as this will only sort the current page being shown.</p>
+
+  <p>Instead, use server-side sorting, with the column headings containing links that reload the page with the table sorted by that heading. You will need to:</p>
```
```suggestion
  <p>Instead, use server-side sorting, with the column header containing links that reload the page with the table sorted by that column. You will need to:</p>
```

anandamaryon1 committed:
fix column/header wording for server-side sorting

Co-authored-by: Frankie Roberto <frankie.roberto1@nhs.net>

anandamaryon1 committed:
Add inital guidance for striped tables

colinrotherham committed:
Remove deprecated `item.header` from responsive tables

colinrotherham committed:
Use updated sortable tables Nunjucks API

colinrotherham committed:
Tables should be sorted in the HTML (no flash of unsorted columns)

anandamaryon1 committed:
Tweak formatting of sortable table examples

anandamaryon1 committed:
tweak guidance for default sort of empty cells + call out  nunjucks options in text

anandamaryon1 committed:
tweak sortable example code

anandamaryon1 committed:
tweak format numeric guidance

colinrotherham commented on app/views/design-system/components/table/index.njk:
```diff
@@ -52,7 +52,8 @@
 
   <h3 id="3-or-more-column-table">3 or more column table</h3>
-  <p>If you have numeric data, right align the column header and cells to make it easier to compare numbers.</p>
+  <p>If you have numeric data, right align the column header and cells to make it easier to compare numbers. If using Nunjucks, set the <code>format: numeric</code> macro option. In HTML add the <code>nhsuk-table__cell--numeric</code> class to the column's header and cells.</p>
```
@anandamaryon1 To complicate things slightly it's:

* `nhsuk-table__cell--numeric` for `<td>` elements
* `nhsuk-table__header--numeric` for `<th>` elements

But take care, `<th>` header cells aren't limited to column headers only

We also use `nhsuk-table__header` in rows with `firstCellIsHeader: true`

anandamaryon1 replied:
Whoops, I'll fix now

colinrotherham commented on app/views/design-system/components/table/index.njk:
```diff
@@ -52,7 +52,8 @@
   <h3 id="3-or-more-column-table">3 or more column table</h3>
-  <p>If you have numeric data, right align the column header and cells to make it easier to compare numbers.</p>
+  <p>If you have numeric data, right align the column header and cells to make it easier to compare numbers. If using Nunjucks, set the <code>format: numeric</code> macro option. In HTML add the <code>nhsuk-table__cell--numeric</code> class to the column's header and cells.</p>
+  <p>For non-numeric data that you want to right align, use the <code>nhsuk-u-text-align-right</code> class instead.</p>
```
Can we better explain how to add classes to cells?

I've seen three different people try to do it this way in recent weeks:

```json5
{
  html: '<span class="nhsuk-u-secondary-text-colour">No data</span>'
}
```

Instead of the correct way:

```json5
{
  text: "No data",
  classes: "nhsuk-u-secondary-text-colour"
}
```

Same advice applies to `nhsuk-u-text-align-right` and other classes

anandamaryon1 replied:
How's this? 
Adding: `If using Nunjucks, set it with <code>classes</code>.`

anandamaryon1 committed:
Explain adding classes via nunjucks option

anandamaryon1 committed:
tweak numeric data guidance

anandamaryon1 committed:
update sorting descriptions based on component updates

colinrotherham commented on app/views/design-system/components/table/sortable-custom/index.njk:
```diff
@@ -0,0 +1,107 @@
+  head: [
+    {
+      text: "Organisation",
+      classes: "nhsuk-u-width-one-third",
```
In the release version we'll have support for the `width` option

```suggestion
      width: "one-third",
```

colinrotherham commented on app/views/design-system/components/table/sortable-custom/index.njk:
```diff
@@ -0,0 +1,107 @@
+    [
+      {
+        text: "Warrington and Halton Hospitals",
+        sortValue: "Warrington and Halton Hospitals"
```
Should we drop the `sortValue` when it's identical? 🤷‍♂️

```suggestion
        text: "Warrington and Halton Hospitals"
```

colinrotherham commented on app/views/design-system/components/table/sortable-custom/index.njk:
```diff
@@ -0,0 +1,107 @@
+      {
+        text: "18 July 2024",
+        sortValue: "2024-07-18",
+        format: "numeric"
```
Are we using `format: "numeric"` in place of `align: "right"` here?

In the sense that `sortValue: "2024-07-18"` is correctly ordered already (numeric not necessary)

```suggestion
        sortValue: "2024-07-18",
        align: "right"
```

anandamaryon1 replied:
Not sure, the `sortVlaue` is a kind of number?

colinrotherham replied:
If you want to, the code doesn't need it though

The ISO 8601 date format (`YYYY-MM-DD`) was ordered that way so A-Z sorting still worked

colinrotherham replied:
If you want to keep it, don't forget the server-side example doesn't use it

(It uses text-align instead of numeric format)

anandamaryon1 replied:
Ok will swap it out when `align: "right"` is ready

colinrotherham commented on app/views/design-system/components/table/sortable-custom/index.njk:
```diff
@@ -0,0 +1,107 @@
+    ],
+    [
+      {
+        text: "The Christie",
+        sortValue: "Christie"
```
They always refer to themselves as "The Christie", so I felt a bit odd about this one

sarawilcox replied:
There are a lot of hospitals that may or may not be searched for under The. Especially when people use their short names

- The Lister
- The Whittington 
- The Royal Free
- The Royal London
- The Royal Marsden
- The Hillingdon Hospitals NHS Foundation Trust
- The Princess Alexandra 
- The Queen Elizabeth Hospital 
- The Royal Orthopaedic 
- The Royal Wolverhampton 
- The Dudley Group 
- The Walton Centre 

to name a few

I think generally speaking it's more helpful to list them under the first letter of the word that follows The.

colinrotherham commented on app/views/design-system/components/table/sortable-custom/index.njk:
```diff
@@ -0,0 +1,107 @@
+      {
+        text: "No data",
+        sortValue: "0",
+        format: "numeric",
```
Due to how [`Intl.Collator` sorting](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Collator) works in browsers we'll need the currency here:

```suggestion
        sortValue: "£0",
        format: "numeric",
```

<br>

Or alternatively you can set `format: "string"` where the `sortValue` is unnecessary

```suggestion
        format: "string",
```

anandamaryon1 replied:
Added the `£`. Not sure what's best though.

anandamaryon1 replied:
Maybe adding the symbols and units is more consistent?

colinrotherham commented on app/views/design-system/components/table/sortable-server-side/index.njk:
```diff
@@ -0,0 +1,188 @@
+    {
+      text: "Last log in",
+      href: "#",
+      classes: "nhsuk-u-text-align-right",
```
```suggestion
      align: "right",
```

anandamaryon1 replied:
Waiting on frontend preview update, will then implement

colinrotherham commented on app/views/design-system/components/table/sortable-server-side/index.njk:
```diff
@@ -0,0 +1,188 @@
+      sort: true
+    },
+    {
+      html: '<span class="nhsuk-u-visually-hidden">Action</span>'
```
Can we make sure classes go in `classes` to remove the nested `<span>`?

Alignment only needs to be set in the header cell, It's auto-applied to the rest of the column

```suggestion
      text: "Action",
      align: "right",
      classes: "nhsuk-u-visually-hidden"
```

anandamaryon1 replied:
Interestingly, adding the class instead of `html` makes an odd gap where the `th` border bottom is missing too. 
![image.png](https://github.com/user-attachments/assets/d65a81af-0f5d-4714-9731-b65f8d5c5fa2)

colinrotherham replied:
Ah because the whole cell is visually hidden

Not great really is it 😬

anandamaryon1 replied:
Have to leave with `html:`?

colinrotherham replied:
If you're still happy with both `item.href` and `item.visuallyHiddenText`?

We can use the latter. I'll open a PR

colinrotherham commented on app/views/design-system/components/table/sortable-server-side/index.njk:
```diff
@@ -0,0 +1,188 @@
+      },
+      {
+        text: "15 January 2026",
+        classes: "nhsuk-u-text-align-right"
```
Will only need `align: "right"` once in the header cell

```suggestion
        text: "15 January 2026"
```

anandamaryon1 replied:
Waiting on frontend preview update, will then implement

colinrotherham commented on app/views/design-system/components/table/sortable-server-side/index.njk:
```diff
@@ -0,0 +1,188 @@
+      },
+      {
+        html: '<a href="#" class="nhsuk-link">Change<span class="nhsuk-u-visually-hidden"> details for Abel Pope</span>',
+        classes: "nhsuk-u-text-align-right"
```
Ahh this is very interesting

Should we allow `item.href` on regular table cells too?

Maybe even `item.visuallyHiddenText`?

```suggestion
        href: "#",
        text: "Change",
        visuallyHiddenText: "details for Abel Pope",
        align: "right"
```

anandamaryon1 replied:
Will wait for `item.href` from frontend

colinrotherham commented on app/views/design-system/components/table/index.njk:
```diff
@@ -25,6 +53,8 @@
 
   <h3 id="3-or-more-column-table">3 or more column table</h3>
   <p>If you have numeric data, right align the column header and cells to make it easier to compare numbers.</p>
+  <p>If using Nunjucks, set the <code>format: numeric</code> macro option. In HTML add the <code>nhsuk-table__header--numeric</code> class to the column's header and <code>nhsuk-table__cell--numeric</code> class to the cells.</p>
```
Can we explain exactly where to set `format: "numeric"`?

* Add Nunjucks `format: "numeric"` to table header cells only
* Add HTML `nhsuk-table__header--numeric` class to table header cells
* Add HTML `nhsuk-table__cell--numeric` class to table body cells

Don't forget the double quotes too:

```suggestion
  <p>If using Nunjucks, set the <code>format: "numeric"</code> macro option. In HTML add the <code>nhsuk-table__header--numeric</code> class to the column's header and <code>nhsuk-table__cell--numeric</code> class to the cells.</p>
```

colinrotherham commented on app/views/design-system/components/table/index.njk:
```diff
@@ -33,14 +63,29 @@
 
   <h2 id="responsive-table">Responsive table</h2>
-  <p>This table has a responsive layout. On large screens it displays as a 3 column table. However, on small screens it stacks vertically.</p>
+  <p>This table has a responsive layout. On large screens it displays horizontally but on small screens it stacks vertically.</p>
```
Can we include instructions on Nunjucks `responsive: true`?

We could explain about the `--responsive` CSS class too but every table element also sets its `role` when responsive (to restore what CSS removes), like `role: "rowheader"` or `role: "cell"`, so it's a bit harder to describe

anandamaryon1 replied:
How's this?

> Enable it in Nunjucks by adding the responsive: true option to the table. In HTML add the nhsuk-table--responsive class to the table and set the correct role attribute for each element of the table.

colinrotherham replied:
Yeah it's better. For reference, we do this in the frontend changelog for complex HTML:

>If you are not using Nunjucks macros, use the HTML markup from the [table examples in the NHS digital service manual](https://service-manual.nhs.uk/design-system/components/table).

colinrotherham commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,14 +95,117 @@
+  <p>Do not use a responsive table if your content is easier to read and understand on a small screen in a non-stacked layout. Use a <a href="#basic-tables">basic table</a> instead. If you have too many columns to fit, use a scrolling table.</p>
+
+  <h2 id="scrolling-table">Scrolling table</h2>
+  <p>Add the scrolling container when your table has many columns and you cannot split it up or use a responsive table.</p>
```
Have we decided what the Nunjucks option should be?

Good to explain that it's the `<div class="nhsuk-table-container"> </div>` wrapper for HTML

anandamaryon1 replied:
`scrollable: true`? Or did you have some thoughts on it?

colinrotherham replied:
Now that we ditched `sortable: true` it can probably go back to non-adjective `scroll: true`?

Otherwise `responsiveable: true` 😉

anandamaryon1 replied:
Ah yes, `scroll: true` is nice, saves the world a few superfluous bytes 👌

anandamaryon1 replied:
Will update once frontend updated and new preview pulled in here

colinrotherham commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,14 +95,117 @@
+  <p>The sortable table component uses JavaScript to sort the data in the user's browser. Use it where you have only 1 page of data. Make sure your data has a suitable sort order if JavaScript does not run.</p>
+
+  <h5 id="default-sort-order">Sorting columns</h5>
+  <p>Decide which column you want to sort by when the page first loads. Make sure users can get back to this. Do not load the page sorted on a column which users cannot sort.</p>
```
What is a column that users cannot sort? Should explain this

colinrotherham commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,14 +95,117 @@
+
+  <h5 id="default-sort-order">Sorting columns</h5>
+  <p>Decide which column you want to sort by when the page first loads. Make sure users can get back to this. Do not load the page sorted on a column which users cannot sort.</p>
+  <p>By default when you click on the column header for an unsorted column, it will sort that column in ascending order. You can change this by adding <code>sortNext: "descending"</code> in Nunjucks (or in HTML <code>data-initial-sort="descending"</code>) to the column header if descending is the more useful way to sort it initially.</p>
```
```suggestion
  <p>By default when you click on the column header for an unsorted column, it will sort that column in ascending order. You can change this by adding <code>sortNext: "descending"</code> in Nunjucks (or in HTML <code>data-sort-next="descending"</code>) to the column header if descending is the more useful way to sort it when clicked.</p>
```

colinrotherham commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,14 +95,117 @@
+  <p>A page refresh or navigation must return the page to its default sort order, unless you add functionality to allow users to save their sort order explicitly.</p>
+
+  <h5 id="sorting-numbers">Sorting numbers</h5>
+  <p>By default, columns will be sorted as text. To correctly order columns containing numbers, set the <code>format: numeric</code> option in Nunjucks. In HTML add the <code>nhsuk-table__header--numeric</code> class to the column header and <code>nhsuk-table__cell--numeric</code> class to the cells).</p>
```
We're kinda [duplicating the content above](https://github.com/nhsuk/nhsuk-service-manual/pull/2382/changes#r3664658961)?

```suggestion
  <p>By default, columns will be sorted as text. To correctly order columns containing numbers, set the <code>format: "numeric"</code> option in Nunjucks. In HTML add the <code>nhsuk-table__header--numeric</code> class to the column header and <code>nhsuk-table__cell--numeric</code> class to the cells).</p>
```

anandamaryon1 replied:
Thoughts @sarawilcox ? 

It's repeated but far up the page. I suppose we can just mention the option? eg.

> By default, columns will be sorted as text. To correctly order columns containing numbers, set the format to numeric. 

Maybe with a link back up?

anandamaryon1 replied:
Decided to keep but reformat a little. Users tend to not read the whole page top to bottom so we feel keeping it is helpful for those looking specifically at sortable tables, and it being a small bit of content, not worth cross linking.

colinrotherham commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,14 +95,117 @@
+    <li>prefixes: The Whittington Hospital would be <code>Whittington Hospital</code></li>
+  </ul>
+
+  <p>Do not sort data alphabetically if it has an established sequence, for example months of the year (January to December).</p>
```
Is this correct?

We can use `"sortValue: "01 January"` to `"sortValue: "12 December"`

anandamaryon1 replied:
But that wouldn't be sorting alphabetically in a user's view. We mean don't sort a column April, August, December etc.

Maybe we should mention months in the custom order section? (but we already mention dates which is similar?)

colinrotherham replied:
Ah but no, see the `01` and `12` prefixes I've added to the `sortValue`?

Users don't see the padded number prefix but it makes sure they stay in order:

* 01 January
* 02 February
* 03 March
* 04 April
* 05 May
* 06 June
* 07 July
* 08 August
* 09 September
* 10 October
* 11 November
* 12 December

Same approach as for dates, works in non-numeric A-Z order too

anandamaryon1 replied:
Removed this line, added a month example to the bullets

colinrotherham commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,14 +95,117 @@
+
+  <p>Do not sort data alphabetically if it has an established sequence, for example months of the year (January to December).</p>
+
+  <p>If a cell contains <a href="#missing-data-table">missing data</a>, say "No data" or similar. By default these cells will be sorted by first letter. With numbers, they will be treated as zero. You can change this by adding a custom <code>sortValue</code>. If using <code>numeric</code> columns with symbols, include the symbol, for example <code>sortValue: "£0"</code>.</p>
```
Probably stick with "alphabetically" instead of "by first letter"?

They won't actually be treated as numbers (sorted numerically) unless `format: "numeric"` is added so it's probably simpler to go straight to `sortValue` at this point:

```suggestion
  <p>If a cell contains <a href="#missing-data-table">missing data</a>, say "No data" or similar. By default these cells will be sorted alphabetically. To sort numerically, add the Nunjucks option <code>sortValue</code> (or in HTML the <code>data-sort-value</code> attribute), for example <code>sortValue: "£0"</code>. Include symbols or units to sort consistently with other cells.
```

colinrotherham commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,14 +95,117 @@
+
+  <h4 id="sort-more-than-1-page-of-data">Sort more than 1 page of data</h4>
+
+  <p>If your table is paginated, do not use the JavaScript-based sorting as this will only sort the current page being shown.</p>
```
We don't split out JavaScript based sorting anymore

```suggestion
  <p>If your table is paginated, sorting should be applied server-side across all pages. Do not sort the current page only.</p>
```

colinrotherham commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,14 +95,117 @@
+
+  <p>If your table is paginated, do not use the JavaScript-based sorting as this will only sort the current page being shown.</p>
+
+  <p>Instead, use server-side sorting, with the column header containing links that reload the page with the table sorted by that column. You will need to:</p>
```
We don't usually say that links reload pages. They navigate elsewhere.

```suggestion
  <p>Use server-side sorting by adding a link to the column header to navigate to a page with the table sorting already applied. You will need to:</p>
```

colinrotherham commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,14 +95,117 @@
+    <li>return users to the first page of new results after sorting</li>
+    <li>test with users</li>
+  </ul>
+  <p>This example shows a paginated table but it does not work in the browser. The functionality would need to be built server-side.</p>
```
Can we please build this?

E.g. We can set up multiple examples and set the `href` to jump between them:

* `sortable-server-side/index.njk`
* `sortable-server-side-ascending/index.njk`
* `sortable-server-side-descending/index.njk`

```suggestion
  <p>This example shows a paginated table with sorting applied server-side.</p>
```

sarawilcox replied:
We have seen 2 people now get confused that this example doesn't actually work. I think we need to say (perhaps in the example itself) that it's not a working example but an illustration of a table with sorting applied server-side.

anandamaryon1 replied:
Ooh nice, yes, I hadn't quite realised it is that simple…

anandamaryon1 replied:
Added new pages and linked them up, so that the example works now. (Used AI, took about 5 tries but saved me some typing and thinking up lots of names)

colinrotherham commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,14 +95,117 @@
+  }) }}
+
+  <h4 id="consider-search-and-filter">Consider search and filter</h3>
+  <p>Sometimes it's better to let users search or filter results instead of sorting. Filtering cuts the data down so users can focus on what they want. Search helps them find specific things they need.</p>
```
Is "cuts the data down" plain English enough?

* When is "sometimes"—when trying to find something?
* Why "instead of sorting"? Use both. E.g. Filter by location, sort by name

```suggestion
  <p>Sometimes it's better to let users search or filter results instead of sorting. Filters reduce the amount of data so that users can focus on what they want. Search helps them find specific things they need.</p>
```

anandamaryon1 replied:
@sarawilcox? 

>When is "sometimes"—when trying to find something?

Yes maybe this can be preceded by: When users need to find a specific thing…?

Think we're trying to say: don't assume sorting will replace the need for filtering and searching.

anandamaryon1 replied:
Applied your rewording and removed first line as it was ambiguous.

colinrotherham commented on app/views/index.njk:
```diff
@@ -86,25 +86,27 @@
           }) }}
         </li>
       </ul>
+      <hr class="nhsuk-u-margin-top-2 nhsuk-u-margin-bottom-5">
```
Can we use a default section break here?

That way we get responsive margins

(Perhaps adjust the bottom margin on the `<ul>` if it's too large)

```suggestion
      <hr class="nhsuk-section-break nhsuk-section-break--l">
```

colinrotherham commented on app/views/index.njk:
```diff
@@ -86,25 +86,27 @@
         </div>
       </div>
-      <hr class="nhsuk-u-margin-top-2 nhsuk-u-margin-bottom-6">
+      <hr class="nhsuk-u-margin-top-3 nhsuk-u-margin-bottom-5">
```
We should use our default section break spacing here too?

```suggestion
      <hr class="nhsuk-section-break nhsuk-section-break--l">
```

anandamaryon1 commented on app/views/design-system/components/table/index.njk:
```diff
@@ -50,14 +95,117 @@
+    group: "components",
+    item: "table",
+    type: "sortable-server-side",
+    showJavaScriptToggle: true
```
I'm thinking we remove this, since it doesn't do anything here. In a sense that's the point, but also, I think it's confusing.

anandamaryon1 committed:
Edits to resolve comments

anandamaryon1 committed:
use responsive margins for hr

anandamaryon1 committed:
remove hr and rows

anandamaryon1 committed:
remove js toggle from server-side example (not needed)

anandamaryon1 committed:
content tweaks to resolve comments

anandamaryon1 committed:
content tweak

anandamaryon1 committed:
Add links to coming soon

colinrotherham commented on app/views/design-system/components/table/index.njk:
```diff
@@ -24,7 +52,12 @@
 
   <h3 id="3-or-more-column-table">3 or more column table</h3>
-  <p>If you have numeric data, right align the column header and cells to make it easier to compare numbers.</p>
+  <p>If you have numeric data, right align the column header and cells to make it easier to compare numbers:</p>
```
Is this line confusing now?

It talks about right aligning the column but we also have our new option now:

* `align: "left"` (default)
* `align: "centre"`
* `align: "right"`

They're shortcuts to add the `nhsuk-u-text-align-*` class

Perhaps the guidance should say that setting the numeric format will automatically right align the column, but alignment can also be set for non-numeric values too? With a link to alignment guidance

anandamaryon1 replied:
It does go on to say about non-numeric, but I need to update it to use `align: "right"` instead of `classes`

anandamaryon1 replied:
How's this?

<img width="755" height="332" alt="image" src="https://github.com/user-attachments/assets/38203139-8cbe-4e13-8d19-fbb43abf2814" />

sarawilcox committed:
Update coming soon page after UR

anandamaryon1 committed:
Mention new align: right option

anandamaryon1 committed:
replace hero image svg

sarawilcox committed:
Updated sortable tables page after UR

anandamaryon1 committed:
tweak hero-image svg code

sarawilcox committed:
Update what's new and change log to include guide to versions

anandamaryon1 committed:
use hero component

anandamaryon1 committed:
Use align-right for date in example

anandamaryon1 committed:
swap scrollable to scroll

anandamaryon1 committed:
add column widths to server side example

colinrotherham committed:
Use new table cell options

sarawilcox committed:
Update What's new and changelog and fix page dates

colinrotherham committed:
Remove unused layouts

colinrotherham committed:
Remove table cell `format: "numeric"` when already set in head

colinrotherham committed:
Add missing macro imports

colinrotherham committed:
Use new heading macro on key pages

colinrotherham committed:
Formatting

sarawilcox committed:
Add link to Coming soon

sarawilcox committed:
Add link to coming soon page

anandamaryon1 committed:
whats new content tweak

sarawilcox committed:
Tweak content after 2i checks

anandamaryon1 committed:
Add missing heading IDs across design system

colinrotherham committed:
Change hero image width to three-quarters

colinrotherham committed:
Update example page spacing to match gutters

colinrotherham committed:
Update childhood vaccination table data

Data source: https://ukhsa-dashboard.data.gov.uk/cover/childhood-vaccinations

Filtered by nation for 6-in-1 (12 months) vaccination

Time series → Tabular data

colinrotherham committed:
Update ibuprofen dosages table data

colinrotherham committed:
Update scroll container guidance

colinrotherham committed:
Try to improve readability of Nunjucks versus HTML options

anandamaryon1 committed:
set summary list values to font weight normal on override classes page

anandamaryon1 committed:
coming soon content tweaks

colinrotherham committed:
Formatting

colinrotherham committed:
Fix tablet specific grid classes description

colinrotherham committed:
Document new conditional display none override classes

colinrotherham committed:
Add viewport widths to override class descriptions

colinrotherham committed:
Fix mixed usage of "above" and "below" for grid widths

colinrotherham committed:
Add reverse table example

colinrotherham committed:
Update NHS.UK frontend preview

colinrotherham committed:
Fix font size override class descriptions to show dual sizes

colinrotherham committed:
Align with other dummy NHS number examples

colinrotherham committed:
Formatting

sarawilcox commented:
This is all looking nice. I made a tiny change to the header for the responsive table. I see that you updated it to match the table on the NHS website page which is great but their table header was confusing me so I've tweaked it. Otherwise it's good that we're following their clinically assured content.

The info about Nunjucks vs HTML is much clearer now. Are we going to be adding the 2 options more widely across the design system? It does add more content to the page. I wonder if we should get Laura to test it in upcoming research?

colinrotherham commented:
>The info about Nunjucks vs HTML is much clearer now. Are we going to be adding the 2 options more widely across the design system? It does add more content to the page. I wonder if we should get Laura to test it in upcoming research?

I've just pushed up https://github.com/nhsuk/nhsuk-service-manual/pull/2382/commits/8a33e31f6a02e441f70e71f7220fdfd161a08e44 to add `variant: "reverse"` guidance

Whether it's the best approach I'm not sure. Our screening teams are widely the same options (but via Jinja)

@anandamaryon1 If you'd like to check the updated options in https://github.com/nhsuk/nhsuk-service-manual/pull/2382/commits/5c0060ca2e8ee71e4f688cf4218ebe7035305f5d please

anandamaryon1 commented:
> I've just pushed up [8a33e31](https://github.com/nhsuk/nhsuk-service-manual/commit/8a33e31f6a02e441f70e71f7220fdfd161a08e44) to add `variant: "reverse"` guidance
> 
> Whether it's the best approach I'm not sure. Our screening teams are widely the same options (but via Jinja)
> 
> @anandamaryon1 If you'd like to check the updated options in [5c0060c](https://github.com/nhsuk/nhsuk-service-manual/commit/5c0060ca2e8ee71e4f688cf4218ebe7035305f5d) please

Thanks, those look good to me @colinrotherham.

I'm not sure on the best approach either. Might be worth some UR as @sarawilcox mentioned. Pros and cons to all approaches… I'll make a ticket to think about it
